File: files/js/src/main/scala/com/swoval/functional/Either.scala
package com.swoval.functional

import Either._
import scala.beans.{ BeanProperty, BooleanBeanProperty }

object Either {

  /**
   * Returns a left projected either
   *
   * @param value The value to wrap
   * @tparam L The type of the left parameter of the result
   * @tparam R The type of the right parameter of the result
   * @tparam T A refinement type that allows us to wrap subtypes of L
   * @return A left projected either
   */
  def left[L, R, T <: L](value: T): Either[L, R] =
    new Left(value.asInstanceOf[L])

  /**
   * Returns a right projected either
   *
   * @param value The value to wrap
   * @tparam L The type of the left parameter of the result
   * @tparam R The type of the right parameter of the result
   * @tparam T A refinement type that allows us to wrap subtypes of R
   * @return A right projected either
   */
  def right[L, R, T <: R](value: T): Either[L, R] =
    new Right(value.asInstanceOf[R])

  class NotLeftException extends RuntimeException

  class NotRightException extends RuntimeException

  class Left[L, R](@BeanProperty val value: L) extends Either[L, R] {

    override def left(): Left[L, R] = this

    override def right(): Right[L, R] = throw new NotRightException()

    override def isLeft(): Boolean = true

    override def isRight(): Boolean = false

    override def toString(): String = "Left(" + value + ")"

    override def equals(other: Any): Boolean = other match {
      case other: com.swoval.functional.Either.Left[_, _] =>
        this.value == other.getValue
      case _ => false

    }

    override def hashCode(): Int = value.hashCode

  }

  class Right[L, R](@BeanProperty val value: R) extends Either[L, R] {

    override def left(): Left[L, R] = throw new NotLeftException()

    override def right(): Right[L, R] = this

    override def isLeft(): Boolean = false

    override def isRight(): Boolean = true

    override def toString(): String = "Right(" + value + ")"

    override def equals(other: Any): Boolean = other match {
      case other: com.swoval.functional.Either.Right[_, _] =>
        this.value == other.getValue
      case _ => false

    }

    override def hashCode(): Int = value.hashCode

  }

}

/**
 * Represents a value that can be one of two types. Inspired by [[https://www.scala-lang.org/api/current/scala/util/Either.html Either]],
 * it is right biased, but does not define all of the combinators that the scala version does.
 *
 * @tparam L The left value
 * @tparam R The right value
 */
abstract class Either[L, R] private () {

  /**
   * Returns the Left projection
   *
   * @return a Left projection
   */
  def left(): Left[L, R]

  /**
   * Returns the Right projection
   *
   * @return a Right projection
   */
  def right(): Right[L, R]

  /**
   * Check whether this is a Left projection.
   *
   * @return true if this is a Reft projection
   */
  def isLeft(): Boolean

  /**
   * Check whether this is a Right projection.
   *
   * @return true if this is a Right projection
   */
  def isRight(): Boolean

  /**
   * Get the right projected value of the either. This is unsafe to call without checking whether
   * the value is a right first.
   *
   * @return the wrapped value if is a right projection
   */
  def get(): R =
    if (isRight) right().getValue else throw new NotRightException()

  /**
   * Get the right projected value of the either or a provided default value.
   *
   * @param r the default value
   * @return the wrapped value if this is a right projection, otherwise the default
   */
  def getOrElse(r: R): R = if (isRight) right().getValue else r

  override def hashCode(): Int

  override def equals(other: Any): Boolean

  /**
   * Casts an either to a more specific left type
   *
   * @param clazz The left type to downcast to
   * @tparam L The original left type
   * @tparam R The right type
   * @tparam T The downcasted left type
   * @return The original either with the left type downcasted to T
   */
  def castLeft[L, R, T <: L](clazz: Class[T]): Either[T, R] =
    if (isRight) this.asInstanceOf[Either[T, R]]
    else if (clazz.isAssignableFrom(left().getValue.getClass))
      this.asInstanceOf[Either[T, R]]
    else
      throw new ClassCastException(left() + " is not an instance of " + clazz)

  /**
   * Casts an either to a more specific right type
   *
   * @param clazz The left type to downcast to
   * @tparam L The original left type
   * @tparam R The right type
   * @tparam T The downcasted right type
   * @return The original either with the right type downcasted to T
   */
  def castRight[L, R, T <: R](clazz: Class[T]): Either[L, T] =
    if (this.isLeft) this.asInstanceOf[Either[L, T]]
    else if (clazz.isAssignableFrom(right().getValue.getClass))
      this.asInstanceOf[Either[L, T]]
    else
      throw new ClassCastException(right() + " is not an instance of " + clazz)

}

File: files/jvm/src/main/java/com/swoval/functional/Either.java
package com.swoval.functional;

/**
 * Represents a value that can be one of two types. Inspired by <a
 * href="https://www.scala-lang.org/api/current/scala/util/Either.html" target="_blank">Either</a>,
 * it is right biased, but does not define all of the combinators that the scala version does.
 *
 * @param <L> The left value
 * @param <R> The right value
 */
public abstract class Either<L, R> {
  private Either() {}

  /**
   * Returns the Left projection for the provided Either or throws an exception if the Either is
   * actually an instance of {@link com.swoval.functional.Either.Right}.
   *
   * @param either the either, assumed to be an instance of left, that will
   * @param <L> the left type of the either.
   * @param <R> the right type of the either.
   * @return a Left projection.
   * @throws NotLeftException if the value is a {@link com.swoval.functional.Either.Right}.
   */
  public static <L, R> Left<L, R> leftProjection(final Either<L, R> either)
      throws NotLeftException {
    if (either.isLeft()) return (Left<L, R>) either;
    else throw new NotLeftException();
  }

  /**
   * Returns the Right projection for the provided Either or throws an exception if the Either is
   * actually an instance of {@link com.swoval.functional.Either.Left}.
   *
   * @param either the either, assumed to be an instance of left, that will
   * @param <L> the left type of the either.
   * @param <R> the right type of the either.
   * @return a Right projection.
   * @throws NotRightException if the value is a {@link com.swoval.functional.Either.Left}.
   */
  public static <L, R> Right<L, R> rightProjection(final Either<L, R> either)
      throws NotRightException {
    if (either.isRight()) return (Right<L, R>) either;
    else throw new NotRightException();
  }

  /**
   * Check whether this is a Left projection.
   *
   * @return true if this is a Reft projection
   */
  public abstract boolean isLeft();

  /**
   * Check whether this is a Right projection.
   *
   * @return true if this is a Right projection
   */
  public abstract boolean isRight();

  /**
   * Get the right projected value of the either. This is unsafe to call without checking whether
   * the value is a right first.
   *
   * @return the wrapped value if is a right projection
   * @throws NotRightException if this is a left projection
   */
  public R get() {
    if (isRight()) return rightProjection(this).getValue();
    else throw new NotRightException();
  }

  /**
   * Get the right projected value of the either or a provided default value.
   *
   * @param either the either from which the method extracts the result if it is a {@link
   *     com.swoval.functional.Either.Right}.
   * @param t the default value
   * @param <T> the default type
   * @return the wrapped value if this is a right projection, otherwise the default
   */
  public static <T> T getOrElse(final Either<?, ? extends T> either, final T t) {
    return either.isRight() ? either.get() : t;
  }

  @Override
  public abstract int hashCode();

  @Override
  public abstract boolean equals(final Object other);

  /**
   * Casts an either to a more specific left type.
   *
   * @param clazz the left type to which we downcast
   * @param <L> the original left type
   * @param <R> the right type
   * @param <T> the downcasted left type
   * @return the original either with the left type downcasted to T.
   * @throws ClassCastException if the wrapped value is not a subtype of T.
   */
  @SuppressWarnings("unchecked")
  public <L, R, T extends L> Either<T, R> castLeft(final Class<T> clazz, final R defaultValue) {
    if (isRight()) {
      return (Either<T, R>) this;
    } else if (clazz.isAssignableFrom(leftProjection(this).getValue().getClass())) {
      return (Either<T, R>) this;
    } else {
      return Either.right(defaultValue);
    }
  }

  /**
   * Casts an either to a more specific right type.
   *
   * @param clazz The right type to which we downcast
   * @param <L> The original left type
   * @param <R> The right type
   * @param <T> The downcasted right type
   * @return The original either with the right type downcasted to T.
   * @throws ClassCastException if the wrapped value is not a subtype of T.
   */
  @SuppressWarnings("unchecked")
  public <L, R, T extends R> Either<L, T> castRight(final Class<T> clazz) {
    if (this.isLeft()) return (Either<L, T>) this;
    else if (clazz.isAssignableFrom(get().getClass())) return (Either<L, T>) this;
    else throw new ClassCastException(rightProjection(this) + " is not an instance of " + clazz);
  }

  /**
   * Returns a left projected either.
   *
   * @param value the value to wrap
   * @param <L> the type of the left parameter of the result
   * @param <R> the type of the right parameter of the result
   * @param <T> a refinement type that allows us to wrap subtypes of L
   * @return A left projected either
   */
  public static <L, R, T extends L> Either<L, R> left(final T value) {
    return new Left<>((L) value);
  }

  /**
   * Returns a right projected either.
   *
   * @param value the value to wrap
   * @param <L> the type of the left parameter of the result
   * @param <R> the type of the right parameter of the result
   * @param <T> a refinement type that allows us to wrap subtypes of R
   * @return a right projected either.
   */
  public static <L, R, T extends R> Either<L, R> right(final T value) {
    return new Right<>((R) value);
  }

  /**
   * An error that is thrown if an attempt is made to project an Either to {@link
   * com.swoval.functional.Either.Left} when the object is actually an instance of {@link
   * com.swoval.functional.Either.Right}.
   */
  public static class NotLeftException extends RuntimeException {}

  /**
   * An error that is thrown if an attempt is made to project an Either to {@link
   * com.swoval.functional.Either.Right} when the object is actually an instance of {@link
   * com.swoval.functional.Either.Left}.
   */
  public static class NotRightException extends RuntimeException {}

  /**
   * A left projected {@link com.swoval.functional.Either}.
   *
   * @param <L> the left type
   * @param <R> the right type
   */
  public static final class Left<L, R> extends Either<L, R> {
    private final L value;

    Left(final L value) {
      this.value = value;
    }

    /**
     * Returns the wrapped value
     *
     * @return the wrapped value
     */
    public L getValue() {
      return value;
    }

    @Override
    public boolean isLeft() {
      return true;
    }

    @Override
    public boolean isRight() {
      return false;
    }

    @Override
    public String toString() {
      return "Left(" + value + ")";
    }

    @Override
    public boolean equals(Object other) {
      return other instanceof com.swoval.functional.Either.Left<?, ?>
          && this.value.equals(((com.swoval.functional.Either.Left<?, ?>) other).getValue());
    }

    @Override
    public int hashCode() {
      return value.hashCode();
    }
  }

  /**
   * A right projected {@link com.swoval.functional.Either}.
   *
   * @param <L> the left type
   * @param <R> the right type
   */
  public static final class Right<L, R> extends Either<L, R> {
    private final R value;

    Right(final R value) {
      this.value = value;
    }

    /**
     * Returns the wrapped value.
     *
     * @return the wrapped value.
     */
    public R getValue() {
      return value;
    }

    @Override
    public boolean isLeft() {
      return false;
    }

    @Override
    public boolean isRight() {
      return true;
    }

    @Override
    public String toString() {
      return "Right(" + value + ")";
    }

    @Override
    public boolean equals(Object other) {
      return other instanceof com.swoval.functional.Either.Right<?, ?>
          && this.value.equals(((com.swoval.functional.Either.Right<?, ?>) other).getValue());
    }

    @Override
    public int hashCode() {
      return value.hashCode();
    }
  }
}
